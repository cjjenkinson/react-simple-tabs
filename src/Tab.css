.tab {
  display: block;
  padding: 0;
  border: 0;
  border-bottom: 2px solid transparent;
  border-radius: 0;
  background: transparent;
  color: #4d4d4d;
  font-size: 14px;
  line-height: 40px;
  letter-spacing: .01em;
  text-align: center;
  text-decoration: none;
  transition: background-color .1s ease-out;
  user-select: none;
  white-space: nowrap;
  cursor: pointer;
}

.tab:hover {
  color: #2B3DBA;
}

.is-active {
  position: relative;
  border-color: #2B3DBA;
  color: #2B3DBA;
  font-weight: 700;
  letter-spacing: 0;
}