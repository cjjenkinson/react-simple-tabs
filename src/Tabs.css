.tabs, .tabs--outer-wrapper {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
}
.tabs--outer-wrapper {
    margin: 0 -15px;
    overflow-x: auto;
}

@media (min-width: 540px) {
  .tabs--outer-wrapper {
    margin: 0 -30px;
  }
}

@media (min-width: 720px) {
  .tabs--outer-wrapper {
      margin: 0;
  }
}

.tabs {
  padding: 0;
  list-style-type: none;
  box-shadow: inset 0 -1px 0 #e3e3e3;
  position: relative;
  flex: 1 1 0%;
}

.tabs > :first-child {
    margin-left: 0;
}

@media (min-width: 960px) {
  .tabs > * {
    margin-left: 15px;
    margin-right: 15px;
  }
}

.tabs > * {
  -webkit-box-flex: 0;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0 10px;
}

.tabs--current-content {
  margin-top: 32px;
}
