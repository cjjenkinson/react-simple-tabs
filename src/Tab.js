import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './Tab.css';

/**
 * A react component that is rendered from Tabs, its children contains the content visible when active.
 * Each Tab passes its index back to Tabs for it to be re-rendered with appropriate active / disabled styles.
 */

class Tab extends Component {
	/**
	 * Function that passes the tab index to the handler within Tabs that determines its
	 * next state e.g active and executes onTabChange callback when passed
	 */
	handleTabChange = e => {
	  const { onClick, tabIndex } = this.props;
	  e.preventDefault();
	  onClick(tabIndex);
	  if (this.props.onTabChange) {
	    this.onTabChangeCallback();
	  }
	};

	/**
	 * Function that invokes the onTabChange function passed as a prop from Tabs
	 */
	onTabChangeCallback = () => this.props.onTabChange();

	/**
	 * Function that renders the Tab with appropriate styling depending on its active state
	 * @param { object } styles an object with the styles applied to the tab markup
	 * @returns { node } markup that will be rendered to the DOM
	 */
	renderTab = () => {
	  const { className, isActive, activeClassName } = this.props;
	  return (
	    <div className={`${className} ${isActive ? activeClassName : ''}`}>
	      <a onClick={this.handleTabChange}>
	        <span>{this.props.title}</span>
	      </a>
	    </div>
	  );
	};

	render() {
	  return this.renderTab();
	}
}

Tab.defaultProps = {
  title: 'Tab Title',
};

Tab.propTypes = {
  onClick: PropTypes.func,
  tabIndex: PropTypes.number,
  isActive: PropTypes.bool,
  onTabChange: PropTypes.func,
  className: PropTypes.string,
  activeClassName: PropTypes.string,
  title: PropTypes.string,
};

export default Tab;
