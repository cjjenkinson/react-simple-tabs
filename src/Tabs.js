import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './Tabs.css';

/**
 * A react component that renders instances of the <Tab /> component as children. When a Tab is changed
 * the Tabs component will keep track of the current tab index and update the content rendered within each Tab.
 */

class Tabs extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentTabIndex: this.props.defaultTabIndex,
    };
  }

	/**
	 * Function that sets the state of the active tab
	 * @param { number } tabIndex index passed from the onClick function
	 */
	handleTabChange = tabIndex => {
	  this.setState({
	    currentTabIndex: this.nextTabIndex(tabIndex),
	  });
	};

	/**
	 * Conditionally checks that the tabIndex is not equal to the current index and updates it accordingly.
	 * @param { number } tabIndex index to check
	 * @returns { number } the next index determined by the current index in state
	 */
	nextTabIndex = tabIndex => {
	  const { currentTabIndex } = this.state;
	  const { defaultTabIndex } = this.props;
	  return tabIndex === currentTabIndex ? defaultTabIndex : tabIndex;
	};

	/**
	 * Function that renders each Tab passing provided props
	 */
	renderTabs() {
	  const { currentTabIndex } = this.state;
	  const { children, onTabChange } = this.props;
	  return React.Children.map(children, (child, index) => {
	    return React.cloneElement(child, {
	      onClick: this.handleTabChange,
	      tabIndex: index,
	      isActive: index === currentTabIndex,
	      onTabChange,
	    });
	  });
	}

	/**
	 * Function that renders the current content from the active Tab
	 */
	renderCurrentTabContent() {
	  const { children } = this.props;
	  const { currentTabIndex } = this.state;
	  if (children[currentTabIndex]) {
	    // render the content inside the instances of <Tab />
	    return children[currentTabIndex].props.children;
	  }
	}

	render() {
	  const { wrapperClassName, className, currentContentClassName } = this.props;
	  return (
	    <div>
	      <div className={wrapperClassName}>
	        <div className={className}>{this.renderTabs()}</div>
	      </div>
	      <section>
	        <div className={currentContentClassName}>
	          {this.renderCurrentTabContent()}
	        </div>
	      </section>
	    </div>
	  );
	}
}

Tabs.defaultProps = {
  defaultTabIndex: 0,
  wrapperClassName: 'tabs--outer-wrapper',
  className: 'tabs',
  currentContentClassName: 'tabs--current-content',
  onTabChange: () => {},
};

Tabs.propTypes = {
  defaultTabIndex: PropTypes.number,
  wrapperClassName: PropTypes.string,
  className: PropTypes.string,
  currentContentClassName: PropTypes.string,
  onTabChange: PropTypes.func,
};

export default Tabs;
