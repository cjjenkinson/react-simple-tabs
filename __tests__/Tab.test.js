import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Tab } from '../src';

import { mount } from 'enzyme';

Enzyme.configure({ adapter: new Adapter() });

describe('React Simple Tabs: Tab', () => {
  it('Should render in the DOM correctly', () => {
    const component = mount(
      <Tab>
        <h1>Billing</h1>
      </Tab>
    );
    expect(component.html()).toMatchSnapshot();
    expect(component.find(Tab).exists()).toBe(true);
  });

  it('should call onClick handler when tab is changed', () => {
    const onClickCb = jest.fn();
    const component = mount(
      <Tab onClick={onClickCb}>
        <h1>Billing</h1>
      </Tab>
    );
    expect(component.html()).toMatchSnapshot();
    component.find('a').simulate('click', { preventDefault: () => {} });
    expect(onClickCb).toHaveBeenCalledWith(component.props().tabIndex);
  });

  it('Should render title passed as props', () => {
    const component = mount(
      <Tab title="Billing" />
    );
    expect(component.html()).toMatchSnapshot();
    expect(component.contains(<span>Billing</span>)).toBeTruthy()
  });

  it('Should invoke onTabChange if passed as props', () => {
    const onClickCb = jest.fn();
    const onTabChangeCb = jest.fn();
    const component = mount(
      <Tab onClick={onClickCb} onTabChange={onTabChangeCb}>
        <h1>Billing</h1>
      </Tab>
    );
    expect(component.html()).toMatchSnapshot();
    component.find('a').simulate('click', { preventDefault: () => {} });
    expect(onClickCb).toHaveBeenCalledWith(component.props().tabIndex);
    expect(onTabChangeCb).toHaveBeenCalled();
  });

  it('Should contain is-active class if active', () => {
    const component = mount(
      <Tab isActive={true} title="Billing">
        <h1>Billing</h1>
      </Tab>
    );
    expect(component.html()).toMatchSnapshot();
    expect(component.find('.is-active').exists()).toBe(true);
  });

  it('Should not contain is-active class if not active', () => {
    const component = mount(
      <Tab isActive={false} title="Billing">
        <h1>Billing</h1>
      </Tab>
    );
    expect(component.html()).toMatchSnapshot();
    expect(component.find('.is-active').exists()).toBe(false);
  });
});
