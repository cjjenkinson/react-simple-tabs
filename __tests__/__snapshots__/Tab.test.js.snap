// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Simple Tabs: Tab Should contain is-active class if active 1`] = `"<div class=\\"tab is-active\\"><a><span>Billing</span></a></div>"`;

exports[`React Simple Tabs: Tab Should invoke onTabChange if passed as props 1`] = `"<div class=\\"tab \\"><a><span>Tab Title</span></a></div>"`;

exports[`React Simple Tabs: Tab Should not contain is-active class if not active 1`] = `"<div class=\\"tab \\"><a><span>Billing</span></a></div>"`;

exports[`React Simple Tabs: Tab Should render in the DOM correctly 1`] = `"<div class=\\"tab \\"><a><span>Tab Title</span></a></div>"`;

exports[`React Simple Tabs: Tab Should render title passed as props 1`] = `"<div class=\\"tab \\"><a><span>Billing</span></a></div>"`;

exports[`React Simple Tabs: Tab should call onClick handler when tab is changed 1`] = `"<div class=\\"tab \\"><a><span>Tab Title</span></a></div>"`;
